graspologic/embed/ase.py

```diff
@@ -158,4 +160,63 @@ def fit(self, graph, y=None):
 
         self.n_features_in_ = len(A)
         self._reduce_dim(A)
+
+        # for out-of-sample
+        inv_eigs = np.diag(1 / self.singular_values_)
+        self.pinv_left_ = self.latent_left_ @ inv_eigs
+        if self.latent_right_ is not None:
+            self.pinv_right_ = self.latent_right_ @ inv_eigs
+
+        self.is_fitted_ = True
         return self
+
+    def predict(self, y):
+        """
+        Obtain out-of-sample latent positions for vertices not in the original embedding.
+        For more details, see [1].
+
+        Parameters
+        ----------
+        y : array_like or tuple, shape (n_oos_vertices, n_vertices)
+            out-of-sample matrix.
+            If tuple, graph is directed and y[0] contains edges from y to other nodes.
```

Comment:
why specifically a tuple as opposed to two variables? seems odd to me.

Reply:
I could set it up that way. My thought was that in the directed case, we want to consider left and right oos vertices as a single object that come in pairs, and so passing it as a tuple made sense to me. 

If we set it up the other way, I'd have to account for the case where the original embedding was directed (e.g., `self.latent_right_` exists), but the user is only passing in one direction for prediction - but maybe we want to be able to do that

Reply:
> If we set it up the other way, I'd have to account for the case where the original embedding was directed - e.g., self.latent_right_ exists, but the user is only passing in one direction for prediction

i dont see why tuple vs two arrays changes this, am I missing something?

> but maybe we want to be able to do that

seems like a niche use case I dont really want to put in extra work to support. cant even think of a reason to do this tbh, if we ever do, can revisit

Reply:
my original thought was that if we do two variables, `self.latent_right_` exists, and the user calls `ase.predict(y)`, then we're in that niche use case and I need to write code to account for that  
if we do tuple, `self.latent_right_` exists, and the user calls `ase.predict(y)`, with `y` being an array, `ValueError` will currently be raised, e.g., there's no way to for the original embedding to be directed but to only pass in one direction.

but I wasn't thinking about the fact that it's only already accounted for because I already wrote code to account for it, lol

I'm happy to do two variables if you want, especially if that's how it's set up in other places in graspologic that take in a directed graph to keep consistency

Reply:
pedigo: fine with tuple but wants sanity check notebook on directions for directed case

Reply:
@bdpedigo @alyakin314 @hhelm10 

documenting here:
in the course of making this notebook, I came across some weirdness in predictions as we increase dimensionality. anton said the 2d part of this notebook happened because the block probability matrix wasn't positive semi-definite. the 3d part of it is also not positive semi-definite.

but predictions were way off with `n_components=3`

https://github.com/loftusa/graspy-personal/blob/master/notebooks/new_latents.ipynb